src/App.tsx

```diff
@@ -1,61 +1,136 @@
+
+  // if there is no user with a given userId
+  return foundUser || null;
+}
+
+const prepTodos: Todo[] = todosFromServer.map(todo => {
+  return {
+    ...todo,
+    user: getUserById(todo.userId),
+  };
+});
 
 export const App = () => {
+  const [title, setTitle] = useState('');
+  const [selectedUser, setSelectedUser] = useState('');
+  const [todos, setTodos] = useState(prepTodos);
+
+  const [shouldErrorOnUserSelect, setErrorOnUserSelect] = useState(false);
+  const [shouldErrorOnTitleInput, setErrorOnTitleInput] = useState(false);
+
+  const handleTitleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    setTitle(event.currentTarget.value);
+    setErrorOnTitleInput(false);
+  };
+
+  const handleUserChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
+    setSelectedUser(event.currentTarget.value);
+    setErrorOnUserSelect(false);
+  };
+
+  const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+
+    setErrorOnTitleInput(!title.trim());
+    setErrorOnUserSelect(!selectedUser);
+
+    if (title.trim() === '' || !selectedUser) {
+      return;
+    }
+
+    const userToAdd = getUserByName(selectedUser);
+
+    setTodos(current => {
+      const maxTodoId = Math.max(...current.map(todo => todo.id));
+
+      return [
+        ...current,
+        {
+          id: maxTodoId + 1,
+          title,
+          completed: false,
+          userId: userToAdd ? userToAdd.id : null,
+          user: userToAdd,
+        },
+      ];
+    });
+
+    setTitle('');
+    setSelectedUser('');
+  };
+
   return (
     <div className="App">
       <h1>Add todo form</h1>
 
-      <form action="/api/users" method="POST">
+      <form
+        action="/api/users"
+        method="POST"
+        onSubmit={handleSubmit}
+      >
         <div className="field">
-          <input type="text" data-cy="titleInput" />
-          <span className="error">Please enter a title</span>
+          <label htmlFor="title"> Write down your task:</label>
+
+          <input
+            type="text"
+            data-cy="titleInput"
+            placeholder="Enter title of task"
+            value={title}
+            onChange={handleTitleChange}
+          />
+          {shouldErrorOnTitleInput && (
+            <span className="error">Please enter a title</span>
+          )}
         </div>
 
         <div className="field">
-          <select data-cy="userSelect">
-            <option value="0" disabled>Choose a user</option>
+          <label htmlFor="title">Choose a user:</label>
+
+          <select
+            data-cy="userSelect"
+            id="userSelect"
+            name="userSelect"
+            value={selectedUser}
+            onChange={handleUserChange}
+          >
+            <option value="" disabled>Choose a user</option>
+
+            {usersFromServer.map(user => (
+              <option key={user.id} value={user.name}>{user.name}</option>
```

Comment:
Let's pass the `id` as value, so you will need a `findById` function
```suggestion
              <option key={user.id} value={user.id}>{user.name}</option>
```